app/services/alerts/generate_and_save_alerts_and_benchmarks.rb

```diff
@@ -80,32 +86,64 @@ def process_alert_benchmark_report(alert_type, alert_report, asof_date)
           )
         end
       else
-        BenchmarkResultError.create!(benchmark_result_school_generation_run: @benchmark_result_school_generation_run, asof_date: asof_date, information: "Relevance: #{alert_report.relevance}", alert_type: alert_type)
+        BenchmarkResultError.create!(benchmark_result_school_generation_run: @benchmark_result_school_generation_run,
+                                     asof_date: asof_date, information: "Relevance: #{alert_report.relevance}", alert_type: alert_type)
       end
     end
 
     def relevant_alert_types
       RelevantAlertTypes.new(@school).list
     end
 
-    def process_alert_type_run_result(alert_type_run_result)
+    def process_alert_type_run_result(alert_type_run_result, alert_attributes: {})
       asof_date = alert_type_run_result.asof_date
       alert_type = alert_type_run_result.alert_type
 
       alert_type_run_result.error_messages.each do |error_message|
-        AlertError.create!(alert_generation_run: @alert_generation_run, asof_date: asof_date, information: error_message, alert_type: alert_type)
+        AlertError.create!(alert_generation_run: @alert_generation_run, asof_date: asof_date,
+                           information: error_message, alert_type: alert_type)
       end
 
       alert_type_run_result.reports.each do |alert_report|
-        process_alert_report(alert_type, alert_report, asof_date)
+        process_alert_report(alert_type, alert_report, asof_date, alert_attributes)
       end
     end
 
-    def process_alert_report(alert_type, alert_report, asof_date)
+    def process_alert_report(alert_type, alert_report, asof_date, alert_attributes)
       if alert_report.valid
-        Alert.create(AlertAttributesFactory.new(@school, alert_report, @alert_generation_run, alert_type, asof_date).generate)
+        Alert.create(AlertAttributesFactory.new(@school, alert_report, @alert_generation_run, alert_type,
+                                                asof_date).generate.merge(**alert_attributes))
       else
-        AlertError.create!(alert_generation_run: @alert_generation_run, asof_date: asof_date, information: "INVALID. Relevance: #{alert_report.relevance}", alert_type: alert_type)
+        AlertError.create!(alert_generation_run: @alert_generation_run, asof_date: asof_date,
+                           information: "INVALID. Relevance: #{alert_report.relevance}", alert_type: alert_type)
+      end
+    end
+
+    def process_custom_periods
+      alert_types = [AlertConfigurablePeriodElectricityComparison, AlertConfigurablePeriodGasComparison,
+                     AlertConfigurablePeriodStorageHeaterComparison].filter_map do |alert_class|
+        AlertType.find_by(class_name: alert_class.name)
+      end
+      Comparison::Report.where.not(custom_period: nil).find_each do |report|
+        alert_types.each do |alert_type|
+          analysis_date = AggregateSchoolService.analysis_date(@aggregate_school, alert_type.fuel_type)
```

Comment:
Testing locally, this is causing `NoMethodError: undefined method amr_data for nil:NilClass` if a school doesn't have the fuel type. Should only be trying to run, e.g. the gas alert if a school `has_gas?`